### PR TITLE
Fix set_subname with exotic names

### PR DESCRIPTION
**Pull request by haarg**
This includes unicode and binary names.  Patches are primarily taken from Sub::Name.

### CODE REVIEW

**Comment by karenetheridge** [comment]
+1

With these changes I would feel more confident in switching Moose from Sub::Name to Sub::Util (see https://github.com/moose/Moose/pull/127)

**Comment by dur-randir** [comment]
Is this any different from #49?

**Comment by haarg** [comment]
I split some of the commits out into separate PRs.

File: ListUtil.xs [hunk 1]
```diff
@@ -1445,14 +1474,18 @@ PPCODE:
...
     CV *cv = NULL;
     GV *gv;
     HV *stash = CopSTASH(PL_curcop);
-    char *s, *end = NULL;
+    const char *s, *end = NULL, *begin = NULL;
     MAGIC *mg;
+    STRLEN namelen;
+    const char* nameptr = SvPV(name, namelen);
+    int utf8flag = SvUTF8(name);
+    int seen_quote = 0, need_subst = 0;
```
Comment (leonerd) [inline ListUtil.xs]: I'm not terribly comfortable with the bookkeeping of `need_subst` and `seen_quote` added in here.

Both variables are declared as `int` types, with names that feel booleanish, but then one of them is used as a counter anyway.

Can you declare these as

    int quotes_seen = 0;
    bool need_subst = FALSE;

Then the various places that `need_subst++` should become

    need_subst = TRUE;

File: ListUtil.xs [hunk 2]
```diff
@@ -1465,63 +1498,81 @@ PPCODE:
...
-        if (*s == ':' && s[-1] == ':')
-            end = ++s;
-        else if (*s && s[-1] == '\'')
-            end = s;
+    for (s = nameptr; s <= nameptr + namelen; s++) {
+        if (s > nameptr && *s == ':' && s[-1] == ':') {
+            end = s - 1;
+            begin = ++s;
+            if (seen_quote)
+                need_subst++;
```
Comment (leonerd) [inline ListUtil.xs]: Why is this line needed? `need_subst` is only used as its truthy nature (see comment above about declaring it `bool`), and the only occasion in which `seen_quote` would be true would have already set `need_subst` true anyway.

File: ListUtil.xs [hunk 3]
```diff
@@ -1465,63 +1498,81 @@ PPCODE:
...
-        char *namepv = savepvn(name, end - name);
-        stash = GvHV(gv_fetchpv(namepv, TRUE, SVt_PVHV));
-        Safefree(namepv);
-        name = end;
+        SV* tmp;
+        if (need_subst) {
+            STRLEN length = end - nameptr + seen_quote - (*end == '\'' ? 1 : 0);
+            char* left;
+            int i, j;
+            tmp = newSV(length);
```
Comment (leonerd) [inline ListUtil.xs]: tmp = sv_2mortal(newSV(length));

would mean you don't have to remember to dec it later

File: ListUtil.xs [hunk 4]
```diff
@@ -1465,63 +1498,81 @@ PPCODE:
...
+                if (nameptr[j] == '\'') {
+                    left[i] = ':';
+                    left[++i] = ':';
+                }
+                else {
+                    left[i] = nameptr[j];
+                }
+            }
+            stash = gv_stashpvn(left, length, GV_ADD | utf8flag);
+            SvREFCNT_dec(tmp);
```
Comment (leonerd) [inline ListUtil.xs]: Unnecessary if you mortalise it earlier

File: ListUtil.xs [hunk 5]
```diff
@@ -1465,63 +1498,81 @@ PPCODE:
...
+        if (s > nameptr && *s == ':' && s[-1] == ':') {
+            end = s - 1;
+            begin = ++s;
+            if (seen_quote)
+                need_subst++;
+        }
+        else if (s > nameptr && *s != '\0' && s[-1] == '\'') {
+            end = s - 1;
+            begin = s;
+            if (seen_quote++)
```
Comment (leonerd) [inline ListUtil.xs]: Are you quite sure you mean `seen_quote++` and not `++seen_quote` here? I.e. the very first time you see a quote mark you increment `seen_quote` but don't yet set `need_subst` to true, until the second time you see one?

File: ListUtil.xs [hunk 6]
```diff
@@ -1465,63 +1498,81 @@ PPCODE:
...
+        SV* old_full_name = newSVpvn_flags(HvNAME(oldhv), HvNAMELEN_get(oldhv), HvNAMEUTF8(oldhv) ? SVf_UTF8 : 0);
+        sv_catpvn(old_full_name, "::", 2);
+        sv_catpvn_flags(old_full_name, GvNAME(oldgv), GvNAMELEN(oldgv), GvNAMEUTF8(oldgv) ? SV_CATUTF8 : SV_CATBYTES);
+
+        old_data = hv_fetch_ent(DBsub, old_full_name, 0, 0);
+
+        SvREFCNT_dec(old_full_name);
+
+        if (old_data && HeVAL(old_data)) {
+            SV* new_full_name = newSVpvn_flags(HvNAME(stash), HvNAMELEN_get(stash), HvNAMEUTF8(stash) ? SVf_UTF8 : 0);
```
Comment (leonerd) [inline ListUtil.xs]: Ditto comment about mortalising here

File: ListUtil.xs (hunk 5 above)
Comment (haarg) [reply ListUtil.xs]: If only a single quote character is seen, and no `::` package separator, then the quote is separating the package name from the sub name.  In that case, we don't need to construct a new package name using double colons (which is what needs_substr tracks) because the package name won't have any single quotes (or `::`) in it.  This also explains why the seen_quote check exists in the branch for `::`.

File: ListUtil.xs (hunk 1 above)
Comment (Leont) [reply ListUtil.xs]: That's probably a historical accident (it's been a while since I touched this, but I think that in an earlier version the two were one variable).

File: ListUtil.xs (hunk 5 above)
Comment (leonerd) [reply ListUtil.xs]: Ah, I see. In that case, maybe rather than two variables could we just count the number of quotes seen (named `quotes_seen` so a plural noun suggesting a count), and change the condition to something like

    if(quotes_seen > 1) {
        /* multiple quotes were seen, but only the final one separates the basename from the package.
           the rest need converting into :: */

Comment (Leont) [reply ListUtil.xs]: It matters where the quote is seen. `Foo::Bar'baz` is fine, but `Foo'Bar::baz` isn't.

Parsing in C sucks.

File: ListUtil.xs (hunk 3 above)
Comment (Leont) [reply ListUtil.xs]: I'd call that a stylistic difference. I don't care much either way, but lets' please keep Sub::Util and Sub::Name in sync :-)

**Comment by haarg** [comment]
@leonerd I've addressed your comments.

**Comment by karenetheridge** [comment]
Do any of the changes here need to get propagated back to Sub::Name?

**Comment by haarg** [comment]
There is no difference in behavior, but if this gets merged I'll prepare a patch for Sub::Name to bring them in line.

**Comment by Leont** [comment]
> if this gets merged I'll prepare a patch for Sub::Name to bring them in line.

@haarg++

**Comment by leonerd** [comment]
It's still a bit messy in places, but I guess that's to be expected of this low level of name mangling. For sake of moving it forward (and that it's consistent with `Sub::Name` anyway), LGTM
